### PR TITLE
Adding Callout Tooltip for Verification Badge tooltip fix#310

### LINKED ISSUE #310: Display tooltip *above*, not over, the verification badge when in Mobile view
from @kevinaboos: 
> in Mobile view, it might be better to display to tooltip above the verification badge instead of directly over it.

[Related PR #306 ](https://github.com/project-robius/robrix/pull/306)

### PR DESCRIPTION
**Pull request by alanpoon**
Fixes #310

1. Added Callout Tooltip for Verification Badge
2. Moved RoomScreenTooltip to shared::callout_tooltip.
3. Passes LiveNodes into Tooltip Actions to style the tooltip

https://github.com/user-attachments/assets/2d5e2258-9734-4310-a8d8-c0839680a80f

### CODE REVIEW

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Thanks for the redesign! I think the new `CalloutTooltip` type is generally a good idea, but I have a few major suggestions for further redesign:

1. The approach of having the `position_helper()` function calculate and return the vector of vectors of live nodes is unnecessary and causes a lot of expensive allocation.
   * Also, because of how Actions work in Makepad, the entire action data will get cloned every time, so we want to keep action data small.
   * Thus, let's not pre-generate these LiveNode double vectors and pass them through action data. It's more expensive and harder to understand. Unless there's a hard reason that we need that, I vote we go back to the simple copy-able types that you had in there before, the widget rect, window size, and width.  The code in `position_helper()` can still be used by / called by the entity that's handling the hover-in action... which I think is similar to what you had before. You could also move that code into a `show_with_options()` function in the CalloutTooltip, which is similar to Alex's original ColorTooltip approach.
       * This is also important to handle weird edge cases where the window size or something else changes in between when the hover-in action is emitted vs when it is handled.
2. If we're replacing ColorTooltip with CalloutTooltip (and I agree that we should), then we ought to remove the ColorTooltip code entirely. CalloutTooltip can offer basic color functionality quite easily.
3. Since we can only ever show a single tooltip at a time, let's just use a single instance in the top-level `App`. Unless there's some reason that we need to have multiple different possible tooltip instances....?
4. Visually, there seems to be a bit too much padding at the bottom here? And the tooltip is really wide.
image

and then in this screenshot, the left padding seem to be a bit too small, in that it's uneven with the top and bottom. I'd also vote to make this tooltip width a bit wider so it's only 2 or 3 lines of text, but that's just my personal opinion.
image

File: src/app.rs [hunk 1]
```diff
@@ -130,7 +131,7 @@ live_design! {
                             verification_modal_inner = <VerificationModal> {}
                         }
                     }
-                    
+                    profile_tooltip = <CalloutTooltip> {}
```
[the remark this answers is not in the document]
Comment (alanpoon) [reply src/app.rs]: fixed

**Comment by alanpoon** [comment]
I have added dynamic tooltip positioning
https://github.com/user-attachments/assets/5aa36a75-0e9e-4798-a679-392405ad9bde

**Review by kevinaboos — CHANGES_REQUESTED** [review]
it looks like you applied some sort of auto-formatter to a lot of code files. Please do not do this in the future, as it makes it really hard for me to properly review your code with a lot of unrelated changes.

Most tooltips are "fine" but they look unprofessional/unpolished. Here are some examples:

The callout triangle isn't properly positioned when the widget being hovered over is at the bottom of the window:
image

Lots of unnecessary extra empty space in the tooltip here:
image
and here:
image

When the window is narrow, the tooltip positioning is completely wrong and off-screen: 
image

And then finally, the video you shared shows a fairly ugly tooltip that is far, far too narrow and tall. Make sure you are selecting reasonable heights/widths for tooltips, such that the text is easy to read and looks polished.
image


In general, your tooltip positioning logic is _way_ over-complicated. Take a look at how I ensure that the right-click context menu is always within the window; the logic is quite simple. Please test things _very_ thoroughly to ensure we're not missing any weird edge cases. We want Robrix to serve as the pinnacle of Rust UI, and as an example of high-quality code.

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Thanks, very nice work! Everything is looking much better. I made some commits for cleanup and to avoid weird interactions between the context menu and the tooltip.

There is one tiny bug remaining, as shown in this screen recording. When the tooltip is first displayed, such as when hovering over a reaction button, it displays as white text with _no_ background for a frame or two. 
(It's a bit hard to see, so you may need to download the video and play it frame-by-frame. But you should be able to easily observe it in Robrix too)


https://github.com/user-attachments/assets/eb15b5c0-8dd9-4691-ae74-306b81bd6ae1

Once you address that, I will merge this in immediately.

**Comment by alanpoon** [comment]
> Thanks, very nice work! Everything is looking much better. I made some commits for cleanup and to avoid weird interactions between the context menu and the tooltip.
> 
> There is one tiny bug remaining, as shown in this screen recording. When the tooltip is first displayed, such as when hovering over a reaction button, it displays as white text with _no_ background for a frame or two. (It's a bit hard to see, so you may need to download the video and play it frame-by-frame. But you should be able to easily observe it in Robrix too)
> 
>  Screen.Recording.2025-02-20.at.2.51.08.PM.mov 
> Once you address that, I will merge this in immediately.

I've fixed it, by hiding the text color if the expected dimension's width is 0.

**Comment by kevinaboos** [comment]
> I've fixed it, by hiding the text color if the expected dimension's width is 0.

oh haha, that's a funny fix. I guess if it works, it works. Thanks.

**Review by kevinaboos — APPROVED** [review]
